extensions/amp-auto-ads/0.1/ad-strategy.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {DataAttributeDef, PlacementState} from './placement';
+import {dev} from '../../../src/log';
+
+/** @const */
+const TAG = 'amp-auto-ads';
+
+export class AdStrategy {
+
+  /**
+   * @param {string} type
+   * @param {!Array<!./placement.Placement>} placements
+   * @param {!Array<!DataAttributeDef>} baseDataAttributes Any data attributes
+   *     that should be added to any inserted ads. These will be combined with
+   *     any additional data atrributes specified by the configuration.
+   */
+  constructor(type, placements, baseDataAttributes) {
+    this.type_ = type;
+
+    this.availablePlacements_ = placements.slice(0);
+
+    this.baseDataAttributes_ = baseDataAttributes;
+
+    /** @type {!Array<!./placement.Placement>} */
+    this.placedAds_ = [];
+  }
+
+  /**
+   * @return {!Promise<boolean>} True if strategy succeeds false otherwise.
+   */
+  run() {
+    // TODO: Implement a more extensive strategy based on parameters set in the
+    // configuration.
+    return this.placeNextAd_();
+  }
+
+  /**
+   * @return {!Promise<boolean>}
+   * @private
+   */
+  placeNextAd_() {
+    const nextPlacement = this.availablePlacements_.shift();
+    if (!nextPlacement) {
+      dev().error(TAG, 'unable to fulfill ad strategy');
```

Comment:
Probably [this](https://github.com/ampproject/amphtml/issues/6417) will be something you need. Still in early design though.

For now, if it's only for local debug, you might want to turn this to warning level. Otherwise, this will show up in AMP's error reporting